New Block: `core/dialog`

sethrubenstein opened the pull request:
## What?
<!-- Link this PR to its associated issue with an appropriate keyword: Closes, See, Follow up to, etc. -->
Closes #61297 

https://github.com/WordPress/gutenberg/issues/61297

This PR introduces 3 blocks from prc-block-library. 

## `core/dialog`

`core/dialog` is actually a collection of three blocks: 1. `core/dialog` a wrapper really for 2. `core/dialog-trigger`, and 3. `core/dialog-element`. 

### `core/dialog-trigger`

This block is the "trigger" or "action" to open the dialog-element `<dialog/>` block. 

### `core/dialog-element`

This is the primary block and is a representation of the `<dialog/>` element in modal form. With added support for positioning center, top left, top center, top right, center left, center right, bottom left, center, and right and support for backdrop coloring it utilize core supports for box shadow and border quite well. 

Additionally, this block and it's iAPI store have been designed to be maximally extensible. We utilize this block throughout pewresearch.org, most complexly in RLS https://www.pewresearch.org/religious-landscape-study/age-distribution/18-29/?dialogId=dialog_prayer-frequency&activeChartId=6ac4c46314ff76e21be70e1f66fe1a19. This link shows off another feature built into `core/dialog`, in this case a modified version of it's built in Deep Linking feature. Which allows you to open a dialog on page render so long as it's id is in the url with `?dialogId`. There is also an Auto Activation timer for opening a dialog immediately on page render based on a ms timer. Lastly, other plugins and consumers of this block can open/close or close all dialogs utilizing the Interactivity API, like so: `store('core/dialog').state.dialogs.[{the dialog's elm id}].isOpen = true` or of course identifying all dialogs on page via `store('core/dialog').state.dialogs`. 

Lastly, there is a block binding provided to associate a heading with the dialogLabel attribute for `core/dialog-element`.
This provides an easy and familiar interface to provide a standard paradigm ~ a heading at the top of a dialog while also allowing the user to set the label and remove the display block, the heading, from the dialog. In either case, if a heading is not present a dynamic heading is created for accessibility and hidden from view on the frontend.

## Why?

Dialogs, or "modals" are a common UI pattern that many 3rd parties have attempted. This provides a base and a representation of an actual HTML element, `<dialog/>` that should be present in the core block library. 

## How?

## Testing Instructions
1. Enable experimental blocks from Gutenberg settings.
2. Open a page/post and insert a Dialog Block
3. Add content to the "trigger"
4. Click "edit dialog" in the block toolbar
5. The dialog should open in the editor
6. Add some content and save the post
7. Preview post and click on your trigger
8. Your dialog should appear

### Testing Instructions for Keyboard

The block toolbar controls should allow keyboard accessibility to open/close dialogs in the editor. And on the frontend the trigger is wrapped with a button element with proper arias to signal relationships. In both contexts there are keyboard handlers to handle escaping out of a dialog.  

## Screenshots or screencast <!-- if applicable -->

https://github.com/user-attachments/assets/4cac53e5-b79b-4349-ac0f-067c412d2b81

https://github.com/user-attachments/assets/d97cd0d2-f79c-47b3-afaf-d9f4190ebd6b

sethrubenstein committed:
Initial commit from prc-block-library, needs to be refactored for core/ namespace

sethrubenstein committed:
Initial commit from prc-block-library, needs to be refactored for core/ namespace

sethrubenstein committed:
Will need to adjust eslint rules to validate :open pseudo element

sethrubenstein committed:
Notes cleanup

sethrubenstein committed:
Code cleanup

sethrubenstein committed:
Cleanup

sethrubenstein committed:
Moving to redux store for editor state management of isOpen

sethrubenstein committed:
Note update

luisherranz commented on packages/block-library/src/dialog-element/view.js:
```diff
@@ -0,0 +1,258 @@
+		get id() {
+			const context = getContext();
+			return context?.id;
+		},
```
You can return a derived state that points to the dialog and then you don't need to make any other derived state (except for `dialogElement`).

```suggestion
		get id() {
			const context = getContext();
			return context?.id;
		},
		get dialog() {
		  return state.dialogs[ state.id ];
		}
```

- ~`get activationTimerDuration()`~ -> `state.dialog.activationTimerDuration`
- ~`get animationDuration()`~ -> `state.dialog.animationDuration `
- ~`get isOpen()`~ -> `state.dialog.isOpen `
- ~`get enableDeepLink()`~ -> `state.dialog.enableDeepLink `
- ...

sethrubenstein replied:
Another great suggestion.

luisherranz commented on packages/block-library/src/dialog-element/view.js:
```diff
@@ -0,0 +1,258 @@
+				}
+				state.dialogs[ id ].isOpen = false;
+
+				// Fire off a custom event to let other blocks know the dialog has closed, it's explicit and only runs when a user closes the dialog.
+				window.dispatchEvent(
+					new CustomEvent( 'wpDialogClosed', { detail: { id } } )
+				);
```
Is there any reason why we need blocks to communicate using events? It's not a pattern we want to promote in the Interactivity API, considering that everything is reactive.

```html
<div data-wp-interactive="other-block" data-wp-watch="callbacks.dialogClosed">...</div>
```

```js
    callbacks: {
        dialogClosed() {
            const { state } = store( 'core/dialog' );
            if ( ! state.dialogs[ id ].isOpen ) {
                // do whatever you need when this dialog is closed...
            }
        }
    }
```

sethrubenstein replied:
This is not necessary. This was just as a hack available to some projects we had not fully converted over to the iAPI yet, but we have since, I'll remove this in my next push. Same for the open event.

luisherranz replied:
Ok, thanks!

luisherranz commented on packages/block-library/src/dialog-element/index.php:
```diff
@@ -0,0 +1,245 @@
+	$enable_deep_link        = array_key_exists( 'enableDeepLink', $attributes ) ? $attributes['enableDeepLink'] : false;
+
+	// By using state any 3rd party can interact as easy as `store('core/dialog').state.dialogs.[blockId].isOpen = true;` which would open the dialog given the blockId.
+	$dialog_state           = wp_interactivity_state(
+		'core/dialog',
+		array()
+	);
+	$dialogs                = $dialog_state['dialogs'] ?? array();
+	$dialogs[ $context_id ] = array(
+		'id'                      => $context_id,
+		'activationTimerDuration' => (int) $auto_activation_timer,
+		'isOpen'                  => $is_open,
+		'enableDeepLink'          => $enable_deep_link,
+		'isClosing'               => false,
+	);
+	wp_interactivity_state(
+		'core/dialog',
+		array(
+			'dialogs' => $dialogs,
+		)
+	);
```
You don't need to do this, associative arrays are not replaced, they are merged.

```suggestion
wp_interactivity_state( 'core/dialog', array( 
	'dialogs' => array(
		$context_id => array(
			'id'                      => $context_id,
			'activationTimerDuration' => (int) $auto_activation_timer,
			'isOpen'                  => $is_open,
			'enableDeepLink'          => $enable_deep_link,
			'isClosing'               => false,
		)
	)
) );
```

sethrubenstein replied:
Oh wow, well I have a ton of block clean up to do then. Much cleaner.

luisherranz commented on packages/block-library/src/dialog-element/view.js:
```diff
@@ -0,0 +1,258 @@
+			) {
+				return;
+			}
+			if ( true !== state.isOpen ) {
```
Maybe if it's already closing `isClosing` as well...?

luisherranz reviewed (commented):
This is awesome, Seth 🙂👏

> Lastly, other plugins and consumers of this block can open/close or close all dialogs utilizing the Interactivity API, like so: `store('core/dialog').state.dialogs.[{the dialog's elm id}].isOpen = true` or of course identifying all dialogs on page via `store('core/dialog').state.dialogs`.

We can combine private and public stores to make this block extensible without exposing all the internal details of the store publicly. It would be the first one from Core, but I think it would be good to do it so we can start exploring extensibility patterns. We can check it a bit later when the new implementation has advanced a bit more.

Other than that, I've only taken a preliminary look at the Interactivity API part, and in general, it's quite good, although here are a few small suggestions.

sethrubenstein committed:
Update packages/block-library/src/dialog-element/index.php

Co-authored-by: Luis Herranz <luisherranz@gmail.com>

sethrubenstein commented:
I'll be making a few more updates by end of the week, mostly around animations and your suggestions @luisherranz.

sethrubenstein commented:
Actually, on further thought, I think I'm going to try to adopt `core/icon` #71227 in some way for the close button. Stay tuned.

sethrubenstein committed:
iAPI cleanup per @luisherranz's recommendations

sethrubenstein committed:
Render effect resolution

sethrubenstein committed:
Cleanup

sethrubenstein committed:
Making store private and creating proxy for read only values, temporary test case for 3rd party interactions

sethrubenstein commented on packages/block-library/src/dialog/store.js:
One thing I wasn't sure about (although I've found it critical to a good experience in the editor) is introducing a custom redux store for a block. I don't believe I've seen a core block do this before. Any thoughts, concerns against this pattern emerging?

sethrubenstein replied:
Going to replace this soon, just familiarizing myself with some recent updates to block attributes around `role:local` which will allow me to use a core method instead of a new redux store to manage this.

sethrubenstein committed:
Critical fix, resolves race condition that can appear in 3rd party usage

sethrubenstein commented:
Once we're done with #69789 I'll turn my attention back to this one and finding alignment with #71227

sethrubenstein commented on packages/block-library/src/dialog-element/style-engine.js:
Based on some feedback from @t-hamano in `core/tabs` on feasibility of maintaining these custom style engine implementations long term I'm rethinking how to handle the backdrop for dialogs. Once I wrap up this tabs refactor I'm going to come back in and create a new `core/dialog-backdrop` block that actually wraps the `core/dialog-element` block. From there background, alignment, justification, etc will make more sense and fall in line with core block use of style supports.

sethrubenstein committed:
Rebase trunk into dialog branch

sethrubenstein committed:
Refactor dialog element block configuration and script registration

- Reformatted JSON structure in block.json for better readability.
- Added 'viewScriptModule' property to block.json for enhanced functionality.
- Removed outdated script registration code from index.php to streamline the rendering process.

sethrubenstein committed:
Add dialog open state management using block attributes

- Introduced `editorIsDialogOpen` attribute to manage the dialog's open state.
- Updated context handling to provide `core/dialog-isDialogOpen` for child blocks.
- Refactored dialog trigger and element components to utilize the new attribute for opening and closing the dialog.
- Removed the previous Redux store implementation for dialog state management, simplifying the architecture.

sethrubenstein committed:
First take at some icons for these blocks.

sethrubenstein committed:
Remove dialog animation, size, and position options. Copied the lightbox animation from core/image into core/dialog-element.

sethrubenstein committed:
Default background and padding for dialog element, open dialog element when selecting the block either in list view or other means.

sethrubenstein committed:
Updates to block removal to trigger when removing a dialog trigger, made it impossible to remove dialog element. Added click handler for backdrop close click in editor.

sethrubenstein committed:
Cleanup labels, sentence case.

sethrubenstein committed:
Try out new naming, the outer block still `core/dialog`, but now its name presents as Dialog Wrapper. `core/dialog-element`, now just named Dialog.

sethrubenstein committed:
Transition to a static block save thats progressively enhanced with iAPI server side instead of a purely server side block. Removes custom style engine implements. This follows the pattern from my work in core/tabs.

sethrubenstein committed:
Fix block removal so it does not fire when we're removing just the dialog wrapper directly.

sethrubenstein committed:
Label updates

sethrubenstein committed:
Yay, the new WP_Icons_Registry is here, and it works! Implemented WP_Icons_Registry for server side close icon in dialog. Notes and style cleanup.

sethrubenstein committed:
Dialog KSES cleanup

sethrubenstein committed:
Cleanup

sethrubenstein committed:
PHPCS fixes

sethrubenstein committed:
Actually, we do need to have animation duration in attributes as it provides the glue between editor and client side interactivity and the animation in style.scss.

This also makes it possible for 3rd parties to change the animation, which pivotaly includes the duration. I agree that including a ui for the editor complicates this but this will allow devs to tap into the block more easily.

sethrubenstein committed:
Cleanup on animation closing timing/logic

sethrubenstein committed:
Just a few last tweaks to some defaults

sethrubenstein committed:
Update dialog element documentation and add description to block.json

sethrubenstein committed:
Description/label clarity

sethrubenstein commented:
I'm considering the image sub-block. It could potentially be interactive with a lightbox now that I think of it. I'll reevaluate some of the sub-block behaviors and clean up any potential interactive elements within the trigger with tag processor tomorrow.

jasmussen commented:
Nice work, I'll start with that. Modal dialogs have by smart folks been described as "the boss battle at the end of the internet". Before I dive into critique I want to acknowledge the monumental task it is to build a fully accessible dialog GUI for 40% of the web. Thanks for the effort.

So please also forgive me as I stumble fresh into a review of this and am finding myself slightly confused at the same time. This is the "new user flow", from someone who've never used this before:

![dialog](https://github.com/user-attachments/assets/96eca906-a00b-4469-814d-82da56f64fbd)

Perhaps evident through the GIF, my confusion becomes clear as soon as I have to decide between Dialog Wrapper and Dialog Label. I think we need better names for these (and by the way I appreciate you added "modal" as metadata so I can search for that too). 

My understanding, and please correct me if I'm wrong, dialogs need accessible screen reader labels. And that if you just add the Dialog Wrapper, you'll get one automatically. And that if you _want_ a screen reader label elsewhere in the DOM, you can add any paragraph or heading block and then use a block binding to connect the two. So does the separate Dialog Label block exist primarily to make it simpler to do this block binding? It seems to just be a heading with a pre-set block binding.

If yes, I'd recommend against that. No other features that use block bindings will be doing it this way, and although it's a nice gesture to simplify, what we really need to do is simplify the block binding UI. If we need a simpler way to provide a custom accessible label for the dialog, possibly something similar to what was done for the Tabs block could be applied here: an input field in the inspector for the Dialog Wrapper block. Which I'd rename to simply "Dialog".

Let me know if that's fair, or if I'm missing nuance!

---

Some suggested description tweaks:

Dialog Wrapper (suggest rename to "Dialog"):

```
Adds a button that opens a modal dialog to display content without leaving the page.
```

Dialog Trigger (suggest rename to "Button"—I know it can be an icon, or a link, but Button may be simpler to understand—perhaps "Trigger Button" if need be):

```
The trigger button that opens the dialog.
```

Dialog (suggest rename to "Content" or "Dialog Content", similar to Accordion block, even using the same icon just as [Tabs](#75195) might):

```
Displays content in a modal dialog that appears on top of the page.
```

---

I'll try and dive into some icons when I have a moment, you've created excellent starting points, and think we can do just small simplicity tweaks to make it all pop, if you'll forgive the pun.

---

For the dialog contents itself:

<img width="569" height="273" alt="image" src="https://github.com/user-attachments/assets/8f19c302-2045-4c1c-9c6e-40e4580c9e71" />

Can we use the `close` icon by default instead of `cancelCircleFilled`? The navigation overlay (CC: @getdave) also uses this icon.

I also love how many block supports you've added. The fact that I can customize the backdrop to be a semi-transparent color of my choice is outstanding. I see I can also add border, shadow, radius:

<img width="566" height="946" alt="image" src="https://github.com/user-attachments/assets/ac2647e1-298e-4a31-bf80-2530fc408fee" />

Those are all hidden by default in the toolspanel, though, which is valid when you want to balance features. However for a dialog specifically, I immediately want to apply a drop shadow and a corner radius, so I think those are worth showing by default.

There's something uncanny about the "Close dialog" button looking toggled here:

<img width="359" height="78" alt="image" src="https://github.com/user-attachments/assets/75cdc6da-28dd-4c92-ad53-dd86234e0925" />

To start, I'd make it look not toggled, if at all possible. 

I appreciate you calling this panel "Settings", because the context of the dialog is sufficient that you don't need a prefix:

<img width="339" height="249" alt="image" src="https://github.com/user-attachments/assets/1b2fad5c-b782-4cf3-8411-5758d62afc6f" />

For the help text:

> Allow the dialog to be opened via a URL hash (e.g., #dialog-id).

Perhaps worth using the term "HTML Anchor" instead of URL hash, if it's not too far a stretch, so we can reuse existing verbiage from here:

<img width="256" height="186" alt="image" src="https://github.com/user-attachments/assets/e535453d-eef4-4271-8a7a-d57537147dec" />

---

The auto activation timer:

<img width="264" height="81" alt="image" src="https://github.com/user-attachments/assets/b8dafc03-11d5-4823-b579-4669781adac8" />

I appreciate the feature-completeness aspect of this block. But I will be candid and note, this feature gives me pause. I know exactly what it does, I see it on every other webpage out there popping up after I've read an article for a bit, prompting me to subscribe to a newsletter. **I don't think we should offer this feature**. It feels like an anti-pattern, perhaps even a dark-pattern. And unless I'm misreading the feature (I trust you all to share nuance if I've missed the point), then I feel this quite strongly. This is not the open web I want to enable through WordPress core. It could/should be a plugin, to be sure, and hopefully the open source code shared in this PR can find its way to a plugin repository if this is desired. But it does not feel like a best practice this project wants to promote. Is that fair?

sethrubenstein commented:
Hi @joedolson not quite yet, hopefully ready for another round of review by Friday. Agreed on the auto activation feature, going to pull that.

sethrubenstein committed:
Making simple style and label updates / pulling auto activation ~ per @jasmussen. Next up, tackling the accessibility issues around dialog trigger...

sethrubenstein committed:
Update dialog-element block namespace to reflect dialog-content

sethrubenstein committed:
Updating attribute and code ownership so dialog prime block owns these things

sethrubenstein committed:
Whoopsies, typo on directive state. Fixes dialog not firing on frontend.

sethrubenstein committed:
Prevent nested interactive elements, namely buttons, from inside the dialog trigger.

sethrubenstein committed:
Refactor dialog block attributes and controls by removing enableDeepLink. Update documentation and improve code formatting for better readability. Improved accessibility in frontend.

sethrubenstein committed:
Fix phpcs linting error

sethrubenstein committed:
Deep link activation cleanup

sethrubenstein committed:
Updates to dialog per @jasmussen

carolinan commented on packages/block-library/src/dialog/readme.md:
```diff
@@ -0,0 +1,7 @@
+
+This implementation includes three blocks: `core/dialog`, `core/dialog-content`, and `core/dialog-trigger`. The `core/dialog` block serves as the container for the dialog, while the `core/dialog-content` block is used as the actual <dialog/> element with styling and content. The `core/dialog-trigger` block is used to create buttons or links that can open the dialog when clicked.
+
+While `core/dialog-trigger` can be removed by the user, `core/dialog-content` is a required inner block of `core/dialog`. This ensures that the dialog has the necessary structure to function properly.
```
I believe this needs to be removed, unless we would like to plan to add readmes to all blocks?

sethrubenstein committed:
Making trigger button ship with core/button only out of the box

sethrubenstein committed:
Updating dialog, getting it into a better state for experimental inclusion more barebones.

sethrubenstein committed:
Cleanup on zoom animation, lock backdrop click state in editor, innerblocks width

carolinan reviewed (changes requested):
Required changes:
The text domain in all block.json file should be "default", not the block slug.

The dialog-trigger has a style.scss file with only a comment, it should be removed.
Both style entries in block.json should be removed.

For the blocks that do have stylesheets, the block.json should refer to the styles with the stylesheet slug like in this example:
```
"editorStyle": "wp-block-cover-editor",
"style": "wp-block-cover"
```


---
Thoughts:
I think using the buttons/button block as the only trigger is working well.

If we want to keep the separate dialog-trigger block instead of making a button variation, in my _opinion_, the block support for the spacing and typography can be removed because it already exists on the buttons block.

Could we have the "Inner blocks use content width" on by default for the dialog content?
Because now when I test the block without it enabled, there is a horizontal scroll.

The dialog content block has default color options visible for buttons. But changing that color does not change the close button.
I'm still wondering if it would not be possible to use the overlay close button:
https://github.com/WordPress/gutenberg/tree/trunk/packages/block-library/src/navigation-overlay-close.
